h/accounts/views.py

```diff
@@ -495,34 +495,49 @@ def get(self):
         """Show the user's account."""
         return self._template_data()
 
-    @view_config(request_method='POST')
-    def post(self):
-        """Handle POST payload from account update form."""
-        formid = self.request.POST.get('__formid__')
-        if formid is None or formid not in self.forms:
-            raise httpexceptions.HTTPBadRequest()
+    @view_config(request_method='POST',
+                 request_param='__formid__=email')
+    def post_email_form(self):
+        """Called by Pyramid when the change email form is submitted."""
+        return self._post(self._handle_email_form)
+
+    @view_config(request_method='POST',
+                 request_param='__formid__=password')
+    def post_password_form(self):
+        """Called by Pyramid when the change password form is submitted."""
+        return self._post(self._handle_password_form)
 
+    def _handle_email_form(self):
         try:
-            if formid == 'email':
-                self._handle_email_form()
-            elif formid == 'password':
-                self._handle_password_form()
+            appstruct = self.forms['email'].validate(self.request.POST.items())
+        except deform.ValidationFailure:
+            raise
```

Comment:
In the case of this particular page it seems to be awkward to use an exception view. The reason is that when a ValidationFailure occurs we need to render two forms:

1. The form that was submitted and failed to validate, with validation errors. If it was the email form that was submitted, for example, then in the view callable `self.forms['email'].render()` will render the email form with validation errors. But in the exception view the same code will render the form _without_ validation errors. Fortunately `self.context.render()` in the exception view will render the submitted form with validation errors, but...

2. We also need to return the other form to the template. This is either `self.forms['email'].render()` or `self.forms['password'].render()` but the exception view doesn't know which. There are ways that the exception view can find out which, or that the main view callable can tell the exception view which, but I think they're all a little awkward.

So I don't think it's possible to _completely_ separate happy path from error handling code, unless this behaviour of deform with exception views can be fixed (I haven't looked into that yet).

So forgetting the exception view for now...

I believe the ValidationFailure error handling as I had it in the _post() method was already generic in that the same code could handle a ValidationFailure from the validation of any form:

```python
try:
    handler()
except deform.ValidationFailure:
    return self._template_data()
```

But there was duplication between the multiple handler methods of the logic to validate the posted data against the form and do the action (change the email or password) if it's valid, this was duplicated in `try ... except ... else ... ` clauses in `_handle_password_form()` and `_handle_email_form()`. I've moved this logic into `_post()` to remove the duplication.

I've also renamed `_post()` to `handle_form_submission()` which I hope is a bit clearer.

This is similar to your `validate_form()` except that instead of returning either an `(appstruct, redirect)` tuple or `(None, template_data)` tuple, it takes an `on_success()` callable as argument. I think this is simpler to use and duplicates less code between the email and password forms.

Of course this does not completely separate happy path from error handling code. Those are slightly mixed together in `handle_form_submission()`. But without the exception view I think this might be the best we can do.